integration-libs/epd-visualization/src/components/visual-picking/visual-picking-product-filter/visual-picking-product-filter.service.ts

```diff
@@ -0,0 +1,67 @@
+import { EventEmitter, Injectable } from '@angular/core';
+import { Product, ProductReference } from '@spartacus/core';
+import { combineLatest, Observable, of, concat } from 'rxjs';
+import { filter, map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class VisualPickingProductFilterService {
+  constructor() {}
+
+  /**
+   * The current filter value.
+   * @param filter The filter value to apply.
+   */
+  public set filter(filter: string) {
+    if (this._filter === filter) {
+      return;
+    }
+    this._filter = filter;
+    this.filter$.emit(filter);
+  }
+  public get filter(): string {
+    return this._filter;
+  }
+  private _filter: string = '';
+  private filter$ = new EventEmitter<string>();
+
+  /**
+   * The set of fields in product objects to perform matching against.
+   */
+  protected fieldsToMatch = ['code', 'name'];
+
+  /**
+   * Returns an Observable that produces a ProductReference[] each time the filter is updated or the set of product references to filter changes.
+   * @param unfilteredProductReferences$ An Observable that returns the unfiltered ProductReference[] to apply filtering to.
+   * @returns An Observable that produces a ProductReference[] each time the filter is updated or the set of product references to filter changes.
+   */
+  public getFilteredProducts$(
+    unfilteredProductReferences$: Observable<ProductReference[]>
+  ): Observable<ProductReference[]> {
+    return combineLatest([
+      concat(of(''), this.filter$),
```

Comment:
why do we need this dummy "" observable here?

Reply:
combineLatest will only return a value when both of its arguments have produced at least one value.
this.filter$ isn't producing anything until the filter value changes, but we want to get results as soon as there are product references available.

I have tried other approaches because this approach bugs me, the alternatives I have tried have had issues (although I could be confusing this case with another similar one). I would like to eliminate this if possible.

Reply:
I have tried this change:

```text
diff --git a/integration-libs/epd-visualization/src/components/visual-picking/visual-picking-product-filter/visual-picking-product-filter.service.ts b/integration-libs/epd-visualization/src/components/visual-picking/visual-picking-product-filter/visual-picking-product-filter.service.ts
index 2011aa729c..61fc33af11 100644
--- a/integration-libs/epd-visualization/src/components/visual-picking/visual-picking-product-filter/visual-picking-product-filter.service.ts
+++ b/integration-libs/epd-visualization/src/components/visual-picking/visual-picking-product-filter/visual-picking-product-filter.service.ts
@@ -1,13 +1,15 @@
 import { EventEmitter, Injectable } from '@angular/core';
 import { Product, ProductReference } from '@spartacus/core';
-import { combineLatest, Observable, of, concat } from 'rxjs';
+import { combineLatest, Observable } from 'rxjs';
 import { filter, map } from 'rxjs/operators';
 
 @Injectable({
   providedIn: 'root',
 })
 export class VisualPickingProductFilterService {
-  constructor() {}
+  constructor() {
+    this.filter$.emit(this.filter);
+  }
 
   /**
    * The current filter value.
@@ -39,10 +41,7 @@ export class VisualPickingProductFilterService {
   public getFilteredProducts$(
     unfilteredProductReferences$: Observable<ProductReference[]>
   ): Observable<ProductReference[]> {
-    return combineLatest([
-      concat(of(''), this.filter$),
-      unfilteredProductReferences$,
-    ]).pipe(
+    return combineLatest([this.filter$, unfilteredProductReferences$]).pipe(
       filter(
         (valuePair: [string, ProductReference[]]) =>
           valuePair[0] !== undefined && valuePair[1] !== undefined
```

But it gives the wrong results. The list that is rendered ends up being initially empty, rather than initially unfiltered. For now I think I will keep the existing code.